app/models/concerns/nested_calculator_validation.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module NestedCalculatorValidation
+  extend ActiveSupport::Concern
+
+  included do
+    validates_associated :calculator
+    after_validation :set_custom_error_messages, if: :calculator_errors?
+  end
+
+  def calculator_errors?
+    calculator.errors.any?
+  end
+
+  def set_custom_error_messages
+    add_custom_error_messages
+    delete_generic_error_message
+    delete_preferred_value_errors
+  end
+
+  def add_custom_error_messages
+    calculator.errors.messages&.each do |attribute, msgs|
+      msgs.each do |msg|
+        errors.add(:base, "#{localize_calculator_attributes[attribute]} #{msg}")
```

Comment:
Why not add the error for the attribute? Maybe because it wouldn't show at the top but we could still include it.
```suggestion
        errors.add(attribute, msg)
```

Reply:
@mkllnk did you mean to add this in addition, or instead of the `:base` error?

Reply:
Instead of. Would that make the error disappear from the top? Can we include attribute errors there, too?

Reply:
Ah I see what you mean now. We add the calculator error to the calculator, and then present calculator errors on-screen along with other errors. I'll give it a go.